mkosi/machine.py

```diff
@@ -122,9 +143,55 @@ def run(self, commands: Sequence[str], timeout: int = 900, check: bool = True) -
 
         return process
 
-    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+    def _kill(self, *args: Any, **kwargs: Any) -> None:
         if self._serial:
             self._serial.kill(signal.SIGTERM)
             self.exit_code = self._serial.wait()
             self._serial = None
         self.stack.__exit__(*args, **kwargs)
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        self._kill(*args, **kwargs)
+
+    def reboot(self) -> None:
+        self._kill(None, None, None)
+        self._boot()
+
+
+class MkosiMachineTest(unittest.TestCase):
+    args: Sequence[str]
+    machine: Machine
+    exc_info: Tuple[Optional[Type[BaseException]], Optional[BaseException], Optional[TracebackType]] \
+              = (None, None, None)
+
+    def _callTestMethod(self, method: str) -> None:
+        try:
+            super()._callTestMethod(method) # type: ignore
+        except Exception as e:
+            self.exc_info = exc_info()
+            raise e
+
+    def __init_subclass__(cls, args: Sequence[str] = []) -> None:
+        cls.args = args
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.machine = Machine(cls.args)
+
+        machine_verb = cls.machine.args.verb
+        no_nspawn = parse_boolean(os.getenv("MKOSI_TESTS_NO_NSPAWN", "0"))
+        no_qemu = parse_boolean(os.getenv("MKOSI_TESTS_NO_QEMU", "0"))
+
+        if no_nspawn and machine_verb == Verb.boot:
+            raise unittest.SkipTest("Nspawn test skipped due to environment variable.")
+        if no_qemu and machine_verb == Verb.qemu:
+            raise unittest.SkipTest("Qemu test skipped due to environment variable.")
+
+        cls.machine.__enter__()
```

Comment:
Given that we'll start booting the machine in setUp(), we should avoid trying to boot the machine here as well.

I'd fix this by adding a public method build() that builds the machine and calling it here instead of `__enter__()`.